8281453: New optimization: convert ~x into -1-x when ~x is used in an arithmetic expression

CptGit opened the pull request:
Similar to `(~x)+c` -> `(c-1)-x` and `~(x+c)` -> `(-c-1)-x` in #6858, we can also introduce similar optimizations for subtraction, `c-(~x)` -> `x+(c+1)` and `~(c-x)` -> `x+(-c-1)`.

To generalize, I convert `~x` into `-1-x` when `~x` is used only in arithmetic expression. For example, `c-(~x)` will be converted into `c-(-1-x)` which will match other pattern and will be transformed again in next iteration and finally become `x+(c+1)`.

Also the conversion from `~x` into `-1-x` happens when `x` is an arithmetic expression itself. For example, `~(x+c)` will be transformed into `-1-(x+c)` and eventually `(-c-1)-x`.

The results of the microbenchmark are as follows:
```
Baseline:                                                               
Benchmark                         Mode  Cnt  Score   Error  Units
NotOpTransformation.baselineInt   avgt   60  0.440 ± 0.002  ns/op
NotOpTransformation.baselineLong  avgt   60  0.440 ± 0.001  ns/op
NotOpTransformation.testInt1      avgt   60  0.613 ± 0.006  ns/op
NotOpTransformation.testInt2      avgt   60  0.868 ± 0.036  ns/op
NotOpTransformation.testLong1     avgt   60  0.674 ± 0.008  ns/op
NotOpTransformation.testLong2     avgt   60  0.698 ± 0.006  ns/op

Patch:
Benchmark                         Mode  Cnt  Score    Error  Units
NotOpTransformation.baselineInt   avgt   60  0.440 ±  0.001  ns/op
NotOpTransformation.baselineLong  avgt   60  0.440 ±  0.001  ns/op
NotOpTransformation.testInt1      avgt   60  0.329 ±  0.001  ns/op
NotOpTransformation.testInt2      avgt   60  0.329 ±  0.001  ns/op
NotOpTransformation.testLong1     avgt   60  0.329 ±  0.001  ns/op
NotOpTransformation.testLong2     avgt   60  0.329 ±  0.001  ns/op

```

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8281453](https://bugs.openjdk.org/browse/JDK-8281453): New optimization: convert ~x into -1-x when ~x is used in an arithmetic expression


### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/7376/head:pull/7376` \
`$ git checkout pull/7376`

Update a local copy of the PR: \
`$ git checkout pull/7376` \
`$ git pull https://git.openjdk.org/jdk pull/7376/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7376`

View PR using the GUI difftool: \
`$ git pr show -t 7376`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/7376.diff">https://git.openjdk.org/jdk/pull/7376.diff</a>

</details>

TobiHartmann commented:
A bug has been filed https://bugs.openjdk.java.net/browse/JDK-8281453.

merykitty commented:
Since `~x == -1 - x` and these 2 operations' costs are essentially the same. It would be much easier if you just check whether the not result is used in an arithmetic operation and transform the former to the latter. The reverse is also true, if you find a `-1 - x` being fed into a bitwise just transform it to a `~x` then.
Thanks.

CptGit commented:
> Since `~x == -1 - x` and these 2 operations' costs are essentially the same. It would be much easier if you just check whether the not result is used in an arithmetic operation and transform the former to the latter. The reverse is also true, if you find a `-1 - x` being fed into a bitwise just transform it to a `~x` then. Thanks.

Thank you for your comment. I am looking at this pending PR again. I drafted it for now until I get a rfr version.

While I am implementing your suggestion, I found another typical optimization that is missing, which is pushing constant to right from `x + (con - y)` to `(x - y) + con`. I created a separate PR for it, https://github.com/openjdk/jdk/pull/7795. Can you take a look at that one if you get a chance? Thank you!

CptGit commented:
Hello @merykitty I finished converting `~x` to `-1-x` when it appears in any add or sub nodes, as you suggested; so now it can really support a wide list of transformation: `c-(~x) => x + (c+1)`, `~x - ~y => y - x`, `(x+1) + ~y => x - y`, etc, which are all created as new tests. Can you help review again thank you.

vnkozlov reviewed (commented):
This change touches the same code as #7795
I suggest to update after that one is pushed.

CptGit commented:
> This change touches the same code as #7795 I suggest to update after that one is pushed.

Merged master after #7795 pushed. Can you look at the updated change? Thank you. @vnkozlov

CptGit commented:
Thanks for reviewing. @vnkozlov 
> Optimization you proposed does not match RFE description and title.

The initial description is somewhat out of date because I adopted @merykitty 's suggestion to include all the cases enabled from `~x => -1-x` instead of adding them one by one, such as `(~x)+c -> (c-1)-x`, which exist in current code base, and many new idealisation enabled from `~x => -1-x` such as `c-(~x) => x + (c+1)`, `~x - ~y => y - x`, `(x+1) + ~y => x - y` I did not update description and title because I'd like to keep history somehow. Please let me know if I should.

> I suggest you to redo performance testing after you merged #7795 changes.

Will redo and post results.

CptGit commented:
/open

CptGit commented:
> Optimization you proposed does not match RFE description and title.
> 
> You do only: `~x` or (x ^ (-1))`->`(-1 - x)`
> 
> As result this should be Xor nodes ideal transformation. I don't even think you need such transformation if `rhs` and `lhs` are not constants because I assume `XOR` and `SUB` hw instructions have the same latency.
> 
> I suggest you to redo performance testing after you merged #7795 changes.

Hello @vnkozlov I updated the performance results and updated description and title as well. Can you please take a look when you get a chance thank you!

vnkozlov reviewed (commented):
This looks much better. You can move PR from Draft state so we can do formal review and testing.

Please, update Subject and Description in JDK-8281453 to match PR's.

Thank you for adding all tests.

CptGit commented:
> Please, update Subject and Description in JDK-8281453 to match PR's.
I have no account on jbs can you help update there thanks!

vnkozlov commented:
> > Please, update Subject and Description in JDK-8281453 to match PR's.
> 
> I have no account on jbs can you help update there thanks!

Done.

vnkozlov reviewed (commented):
I submitted testing.

vnkozlov reviewed (approved):
Testing passed.

You need second review.

CptGit commented:
> Testing passed.
> 
> You need second review.

Thank you @vnkozlov for approving. Do you have some people for reviewing suggesting for me?

vnkozlov commented:
@merykitty can review it since he gave initial comment.

merykitty commented:
Hi, what if `~x` is also used in another context, would it duplicate this local, also how about the cases where `x` is an addition or subtraction? I think this should be an idealisation of the `XorNode`, and if you cannot really check the uses of a node during parsing, you can record them for igvn later. Thanks.

CptGit commented:
Thanks @merykitty  for the comment. Can you give an example for this?
>what if `~x` is also used in another context, would it duplicate this local

Also, I think this works.
> how about the cases where `x` is an addition or subtraction

merykitty commented:
Consider there are 2 statements:

    int x = ~a + b;
    int y = ~a | c;

Your transformation would duplicate the `~a` as it is transformed into `(-a - 1)` in the calculation of `x` but not in the calculation of `y`.

Thanks.

CptGit commented:
Thanks @merykitty  So do you mean we want to do the transformation only when ~a has one use, i.e., `outcnt() == 1`? This makes sense to me.

However I still don't get why it should be done in `XOrNode`. Could you shed light on this?

> Consider there are 2 statements:
> 
> ```
> int x = ~a + b;
> int y = ~a | c;
> ```
> 
> Your transformation would duplicate the `~a` as it is transformed into `(-a - 1)` in the calculation of `x` but not in the calculation of `y`.
> 
> Thanks.

merykitty commented:
> So do you mean we want to do the transformation only when ~a has one use, i.e., `outcnt() == 1`

Yes, you can iterate through the uses of the node to be more general

> However I still don't get why it should be done in XOrNode

Because this seems logically to be a transformation of the not node, not of the add or sub node, and putting the transformation there may hurt maintainability.

Thanks.

CptGit commented:
@merykitty Sounds good! I'm implementing this in XorNode.

I came a question on `outcnt()`, I included in `XorINode::Ideal`
```
if (phase->type(in2) == TypeInt::MINUS_1) {
  std::cerr << "!!!outcnt() = " << outcnt() << std::endl;
  if (outcnt() == 1) {
    int user_op = unique_out()->Opcode();
    if (user_op == Op_AddI || user_op == Op_SubI) {
      return new SubINode(in2, in1);
    }
  }
}
```
but I found tests failed, for example
```
@Test
@IR(failOn = {IRNode.SUB, IRNode.XOR, IRNode.ADD})
// Checks -1 - ~x => x
public int test25(int x) {
    return -1 - ~x;
}
```
After some debugging, I found the cause was `outcnt()` is `0` so the transformation did not happen at all (the IR printed stills follows the original structure of `-1-~x`). I don't understand why `outcnt()` is `0` not `1` since `~x` is used by `-1 - ~x`. Appreciate it if you can shed some light on this. Thank you.

merykitty commented:
@CptGit It is due to the fact that GVN runs as soon as the frontend parses the code, during which the graph is incomplete, and you get `outcnt() == 0` because the uses of the node have not been parsed yet. You can defer the transformation to IGVN, which happens later. Thanks.

CptGit commented:
@merykitty I moved things to `XorINode` and `XorLNode` and made generalization. Can you please have a look when you get a chance if my changes make sense? Thank you.

merykitty commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -872,20 +863,38 @@ const Type *OrLNode::add_ring( const Type *t0, const Type *t1 ) const {
+    if (phase->is_IterGVN()) {
+      if (is_used_in_only_arithmetic(this, T_INT)
+          // used only once and contains arithmetic
+          || (outcnt() == 1 && (in1->Opcode() == Op_AddI || in1->Opcode() == Op_SubI))) {
```
May I ask why do you need `outcnt() == 1` here, thanks.

CptGit replied:
I was thinking of which cases we still want to transform even when the `Xornode` is used in a non-arithmetic expression, and the only case I think safe to transform is that when it is used only once and has some arithmetic inside, for example,
`return ~(x+c)`.

merykitty replied:
In that case, you would need to check the `Opcode` of the use similar to the above line. However, I believe that since you have already deferred the transformation, the bitwise transformations should have been finished by then so you don't need to worry about that. I would suggest adding a comment to remind this. And some tests to confirm, such as `(x + y) & ~(x + y) == 0`. Thanks.

CptGit replied:
Thanks. I agree transformation like `(x + y) & ~(x + y) => 0` should have been covered by idealization from `AndINode` at this point. However, my intention was to catch the case like `return ~(x-1) => return -x`. If we have only line 889 we would not be able to do this because `~(x - 1)` is not used in arithmetic and will not be transformed.

merykitty replied:
I mean you don't need `outcnt() == 1`.

CptGit commented:
@merykitty I removed the use check. Does it look good to you? I did not include test `(x + y) & ~(x + y) => 0` because I found we do not have such idealization in `AndINode` because even `x & ~x => 0` is not supported.

merykitty commented:
I think it looks good, thanks.

vnkozlov reviewed (commented):
I will test latest version.

vnkozlov reviewed (approved):
Latest testing passed.

vnkozlov commented:
@CptGit  Did you updated performance numbers for latest changes? Did they change?

CptGit commented:
> @CptGit Did you updated performance numbers for latest changes? Did they change?

@vnkozlov Updated now. Thanks.

vnkozlov commented:
Good. It is ready for integration.

CptGit commented:
/integrate

vnkozlov commented:
> @CptGit This pull request has not yet been marked as ready for integration.

I think, it is because PR title does not match JBS entry title.

CptGit commented:
/integrate

vnkozlov commented:
/sponsor
